lib/client/weblogin.go

```diff
@@ -658,3 +673,248 @@ func GetWebConfig(ctx context.Context, proxyAddr string, insecure bool) (*webcli
+			return nil, err
+		}
+
+		req, err := http.NewRequestWithContext(ctx, "POST", clt.Endpoint("webapi", "sessions", "web"), &buf)
+		if err != nil {
+			return nil, err
+		}
+
+		cookie := &http.Cookie{
+			Name:  csrf.CookieName,
+			Value: csrfToken,
+		}
+
+		req.AddCookie(cookie)
+
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set(csrf.HeaderName, csrfToken)
+		return clt.HTTPClient().Do(req)
+	}))
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	session, err := GetSessionFromResponse(resp)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return clt, session, nil
+}
+
+// SSHAgentMFAWebSessionLogin requests a MFA challenge via the proxy web api.
+// If the credentials are valid, the proxy will return a challenge. We then
+// prompt the user to provide 2nd factor and pass the response to the proxy.
+func SSHAgentMFAWebSessionLogin(ctx context.Context, login SSHLoginMFA) (*WebClient, types.WebSession, error) {
+	clt, _, err := initClient(login.ProxyAddr, login.Insecure, login.Pool, login.ExtraHeaders)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	beginReq := MFAChallengeRequest{
+		User: login.User,
+		Pass: login.Password,
+	}
+	challengeJSON, err := clt.PostJSON(ctx, clt.Endpoint("webapi", "mfa", "login", "begin"), beginReq)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	challenge := &MFAAuthenticateChallenge{}
+	if err := json.Unmarshal(challengeJSON.Bytes(), challenge); err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// Convert to auth gRPC proto challenge.
+	challengePB := &proto.MFAAuthenticateChallenge{}
+	if challenge.TOTPChallenge {
+		challengePB.TOTP = &proto.TOTPChallenge{}
+	}
+	if challenge.WebauthnChallenge != nil {
+		challengePB.WebauthnChallenge = wanlib.CredentialAssertionToProto(challenge.WebauthnChallenge)
+	}
+
+	respPB, err := PromptMFAChallenge(ctx, challengePB, login.ProxyAddr, &PromptMFAChallengeOpts{
+		AllowStdinHijack:        login.AllowStdinHijack,
+		AuthenticatorAttachment: login.AuthenticatorAttachment,
+		PreferOTP:               login.PreferOTP,
+	})
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	challengeResp := AuthenticateWebUserRequest{
+		User: login.User,
+	}
+	// Convert back from auth gRPC proto response.
+	switch r := respPB.Response.(type) {
+	case *proto.MFAAuthenticateResponse_Webauthn:
+		challengeResp.WebauthnAssertionResponse = wanlib.CredentialAssertionResponseFromProto(r.Webauthn)
+	default:
+		// No challenge was sent, so we send back just username/password.
+	}
+
+	loginRespJSON, err := clt.PostJSON(ctx, clt.Endpoint("webapi", "mfa", "login", "finishsession"), challengeResp)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	session, err := GetSessionFromResponse(loginRespJSON)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	return clt, session, nil
+}
+
+// SSHAgentPasswordlessLoginWeb requests a passwordless MFA challenge via the proxy.
+// web api.
```

Comment:
It looks like "proxy" in the comment was supposed to be replaced with "web api".